src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp

```diff
@@ -16,112 +16,144 @@
-                BansheeCurse_Timer = 18000;
-            } else BansheeCurse_Timer -= diff;
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
 
-            //Silence
-            if (Silence_Timer <= diff)
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
             {
-                if (rand32() % 100 < 80)
-                    DoCastVictim(SPELL_SILENCE);
-                //13 seconds until we should cast this again
-                Silence_Timer = 13000;
-            } else Silence_Timer -= diff;
-
-            DoMeleeAttackIfReady();
+            case EVENT_SPELL_BANSHEEWAIL:
+                DoCastVictim(SPELL_BANSHEEWAIL);
+                _events.ScheduleEvent(EVENT_SPELL_BANSHEEWAIL, 4s);
+                break;
+            case EVENT_SPELL_BANSHEECURSE:
+                DoCastVictim(SPELL_BANSHEECURSE);
+                _events.Repeat(18s);
+                break;
+            case EVENT_SPELL_SILENCE:
+                DoCastVictim(SPELL_SILENCE);
+                _events.ScheduleEvent(EVENT_SPELL_SILENCE, 13s);
+                break;
+            case EVENT_SPELL_POSSESS:
+                if (Unit* possessTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false))    // Random target to be possessed
+                {
+                    if (possessTarget->IsAlive())
+                    {
+                        me->CastStop();
+                        DoCast(possessTarget, SPELL_POSSESS);
+                        DoCast(possessTarget, SPELL_POSSESSED);
+                        me->SetReactState(REACT_PASSIVE);
+                        me->SetVisible(false);
+                        _possessedTargetGuid = possessTarget->GetGUID();
+                        _events.ScheduleEvent(EVENT_CHECK_POSSESSED, 0s);
+                    }
+                    else
+                        _events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+                }
+                break;
+            case EVENT_INVISIBLE:
+                if (Player* possessedTarget = ObjectAccessor::FindConnectedPlayer(_possessedTargetGuid))    // When there's a possessed target
+                {
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESS);
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESSED);
+                    me->SetVisible(true);
+                    possessedTarget->SetFullHealth();
+                    me->SetReactState(REACT_AGGRESSIVE);
+                    _events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+                }
+                break;
+            case EVENT_CHECK_POSSESSED:
+                if (!me->IsVisible())
+                {
+                    if (Player* possessedTarget = ObjectAccessor::FindConnectedPlayer(_possessedTargetGuid))
+                    {
+                        if (!possessedTarget->HasAura(SPELL_POSSESSED) || possessedTarget->GetHealthPct() <= 50)
```

Comment:
I think there is a helper for checking hp <= pct.... `IsHealthBelow` something or another I think?

Reply:
also this whole event feels hacky

is there maybe a dummy periodic on the possess aura that you could attach a spellscript to or something?